csharp/src/Ice/Reference.cs

```diff
@@ -584,6 +584,101 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
+                        if (IsConnectionCached)
+                        {
+                            _connection = connection;
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        if (i + 1 == connectors.Count)
+                        {
+                            throw;
+                        }
+                    }
+                }
+            }
+            Debug.Assert(connection != null);
+            return connection;
+        }
+
+        internal async ValueTask<(Connection? Connection, bool Cached, IReadOnlyList<Endpoint> Endpoints, List<Connector>? Connectors)> GetExistingConnectionAsync(CancellationToken cancel)
+        {
+            Connection? connection = GetCachedConnection();
+            bool cached = false;
+            IReadOnlyList<Endpoint>? endpoints = ImmutableList<Endpoint>.Empty;
+            List<Connector>? connectors = null;
+
+            if (connection == null || (!IsFixed && !connection.IsActive))
+            {
+                // No cached connection, so now check if there's a connection to one of our endpoints if the
+                // prefer existing connection property is true.
+
+                // TODO replace IsConnectionCached with PreferExistingConnection
+                if (IsConnectionCached)
+                {
+                    // Compute the endpoints (eventually performing a locator lookup, moving last the recently
+                    // failed endpoints, filtering endpoints based on the proxy properties). If the endpoints
+                    // are retrieved from a cache, a cached serial is returned and identifies the lookup of the
+                    // cached entry. This is useful to ensure we don't perform a new lookup if the cached entry
+                    // has already been updated since we got it.
+                    (cached, endpoints) = await ComputeEndpointsAsync(cancel).ConfigureAwait(false);
+
+                    // The communicator returns an existing  connection if one is already established to one of the given
+                    // endpoints.
+                    OutgoingConnectionFactory connectionFactory = Communicator.OutgoingConnectionFactory;
+                    connection = connectionFactory.GetConnection(endpoints, ConnectionId);
+
+                    // If we couldn't find the connection using the endpoints, expand the endpoints to
+                    // connectors (DNS lookup for IP endpoints) and try to obtain a connection using the
+                    // connectors.
+                    if (connection == null)
+                    {
+                        // This will throw if no connectors can be computed.
+
+                        // TODO retry with fresh endpoints if ComputeConnectorsAsync throws and we are using cached
+                        // endpoints.
```

Comment:
Yes but I don't think this retry belongs here. It's up to the caller to retry if he decides to.